pmd-core/src/main/java/net/sourceforge/pmd/util/IOUtil.java

```diff
@@ -120,7 +120,21 @@ public static Writer createWriter(Charset charset, String reportFile) {
                 return new OutputStreamWriter(new FilterOutputStream(System.out) {
                     @Override
                     public void close() {
-                        // do nothing, avoid closing stdout
+                        // avoid closing stdout, simply flush
+                        try {
+                            out.flush();
+                        } catch (IOException ignored) {
+                            // Nothing left to do
+                        }
+                    }
+                    
+                    @Override
+                    public void write(byte[] b, int off, int len) throws IOException {
+                        /*
+                         * FilterOutputStream iterates over each byte, asking subclasses to provide more efficient implementations
+                         * It therefore negates any such optimizations that the underlying stream actually may implement.
+                         */
+                        out.write(b, off, len);
```

Comment:
This was extremely unexpected. `FilterOutputStream.write` is the only `write` method that don't delegate to the wrapper `OutputStream`, but manually iterates and calls `write(int)` with the Javadoc explicitly stating this and recommending:

> Note that this method does not call the {@code write} method
     of its underlying output stream with the same arguments. Subclasses
     of {@code FilterOutputStream} should provide a more efficient
     implementation of this method.

I do not understand the rationale behind this. This is not only slower, but also introduces a bunch of race conditions, as `PrintStream` (ie: `System.out`) synchronizes all write operations, so messages can't interleave. This thread safety is lost when each char is written on a separate call.

While doing so, I also realized we were missing a forceful flush upon "closing".

Reply:
Interesting... I suppose it might be because FilterOutputStream's base implementation is made for byte-by-byte filtering, and it would be inconsistent for it not to implement the the bulk write method in terms of its other write method (would force all implementors to implement both methods)... Anyway nice find

Reply:
No, that can't be it… otherwise they would have made the same for `FilterOutputStream.write(byte[] buf)`, but they didn't. No clue really, the more I think about it, the less sense it makes 🤷